pallets/sora2-trustless-bridge/beefy-light-client/src/lib.rs

```diff
@@ -0,0 +1,587 @@
+// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use bridge_common::{beefy_types::*, bitfield};
+use codec::Encode;
+use frame_support::log;
+use frame_support::traits::Randomness;
+pub use pallet::*;
+use scale_info::prelude::vec::Vec;
+use sp_io::hashing::keccak_256;
+use sp_core::{H256};
+
+pub use bitfield::BitField;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+fn recover_signature(sig: &[u8; 65], msg_hash: &[u8; 32]) -> Option<EthAddress> {
+    use sp_io::{
+        crypto::secp256k1_ecdsa_recover,
+    };
+
+	secp256k1_ecdsa_recover(sig, msg_hash)
+		.map(|pubkey| EthAddress::from(H256::from_slice(&keccak_256(&pubkey))))
+		.ok()
+}
+
+impl<T: Config> Randomness<sp_core::H256, T::BlockNumber> for Pallet<T> {
+    fn random(_: &[u8]) -> (sp_core::H256, T::BlockNumber) {
+        let block_number = <frame_system::Pallet<T>>::block_number();
+        (Self::get_seed().into(), block_number)
+    }
+
+    fn random_seed() -> (sp_core::H256, T::BlockNumber) {
+        Self::random(&[][..])
+    }
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+    use bridge_common::{merkle_proof, simplified_mmr_proof::*};
+    use ethabi::{encode_packed, Token};
+    use frame_support::fail;
+    use frame_support::pallet_prelude::OptionQuery;
+    use frame_support::{dispatch::DispatchResultWithPostInfo, pallet_prelude::*};
+    use frame_system::pallet_prelude::*;
+
+    pub const MMR_ROOT_HISTORY_SIZE: u32 = 30;
+    pub const THRESHOLD_NUMERATOR: u128 = 22;
+    pub const THRESHOLD_DENOMINATOR: u128 = 59;
+    pub const NUMBER_OF_BLOCKS_PER_SESSION: u64 = 600;
+    pub const ERROR_AND_SAFETY_BUFFER: u64 = 10;
+    pub const MAXIMUM_BLOCK_GAP: u64 = NUMBER_OF_BLOCKS_PER_SESSION - ERROR_AND_SAFETY_BUFFER;
+    pub const MMR_ROOT_ID: [u8; 2] = [0x6d, 0x68];
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+        type Randomness: frame_support::traits::Randomness<Self::Hash, Self::BlockNumber>;
+    }
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    #[pallet::without_storage_info]
+    pub struct Pallet<T>(_);
+
+    // The pallet's runtime storage items.
+    #[pallet::storage]
+    #[pallet::getter(fn latest_mmr_roots)]
+    pub type LatestMMRRoots<T> = StorageMap<_, Blake2_256, u128, [u8; 32], ValueQuery>;
```

Comment:
I think it's better to use H256 instead of [u8; 32] here and in another places, because it's more readable and also have hex formatting by default

Reply:
I think it's better to use StorageValue with VecDeque, because it will decrease storage reads and will be easier to maintain and understand